src/Tokenizer.php

```diff
@@ -766,34 +766,17 @@ public function tokenize(string $string): Cursor
     {
         $tokens = [];
 
-        // Used to make sure the string keeps shrinking on each iteration
-        $oldStringLen = strlen($string) + 1;
-
         $token = null;
 
-        $currentLength = strlen($string);
-
         // Keep processing the string until it is empty
-        while ($currentLength) {
-            // If the string stopped shrinking, there was a problem
-            if ($oldStringLen <= $currentLength) {
-                $tokens[] = new Token(Token::TOKEN_TYPE_ERROR, $string);
-
-                return new Cursor($tokens);
-            }
-
-            $oldStringLen =  $currentLength;
-
+        while ($string !== '') {
             // Get the next token and the token type
-            $token       = $this->createNextToken($string, $token);
-            $tokenLength = strlen($token->value());
+            $token = $this->createNextToken($string, $token);
 
             $tokens[] = $token;
 
             // Advance the string
-            $string = substr($string, $tokenLength);
-
-            $currentLength -= $tokenLength;
+            $string = substr($string, strlen($token->value()));
```

Comment:
the length check was done to ensure `$token->value()` consumed at least one chr... if for some reason did not consume any chr we would have an infinite loop. 

I trust that `createNextToken` will consume at least one chr... but im not 100% sure it will be always so as it is quite complex

Reply:
@goetas In token constructor we assert that the value is non empty string, thus we can rely on that safely and `strlen($token->value()) === 0` will never be zero.